### ISSUE TITLE
getProfile() will start breaking soon

### ISSUE THREAD

**Issue by glacambre** [tridactyl/tridactyl#1627]
Mozilla is changing the way profiles are handled with newer firefox versions. Users will get both an `installs.ini` file and a `profiles.ini` file and we will need to consult both these files in order to figure out what the default profile is.

**Comment by Rummskartoffel** [comment]
According to [the support article](https://support.mozilla.org/en-US/kb/understanding-depth-profile-installation) about this change, `installs.ini` is only used to fix `profiles.ini` in case an old version of Firefox breaks it. Since Tridactyl only ever runs in Firefox versions that don't break `profiles.ini` (and fix it, should they find it broken), we should be able to disregard `installs.ini`.

**Comment by Rummskartoffel** [comment]
I started looking at this again. To know the correct default profile, we need to find out which install we are running on in case there are multiple. 
On Windows, we can do that by reading `HKCU\Software\Mozilla\Firefox\TaskBarIDs\<Firefox install directory>`, which helpfully stores the hash we need. For `C:\Program Files\Mozilla Firefox`, for instance, that hash is `308046B0AF4A39CB`, so we need to read the `Default` key of section `[Install308046B0AF4A39CB]` in `profiles.ini`.
On Linux, that hash doesn't seem to be stored anywhere – or at least I couldn't find it using `find / -xdev -type f -print0 2>/dev/null | xargs -0 grep -H "<hash, see profiles.ini>"` – , so we'd have to generate it ourselves from the the directory the `firefox` binary is located in using [`CityHash64`](https://github.com/google/cityhash), which sounds like it would be wildly impractical.

**Comment by bovine3dom** [comment]
You could spelunk the Firefox codebase with searchfox - https://searchfox.org/ - to work out where profiles.ini is read and therefore how the right section is found across all platforms

**Comment by bovine3dom** [comment]
My phone's web browser can't handle the page but https://searchfox.org/mozilla-central/source/toolkit/profile/nsToolkitProfileService.h already looks promising : )

**Comment by Rummskartoffel** [comment]
I did a bit of digging through there already – that's how I found out what I found out. I hadn't found that file yet, though. Here's what I found: 

* [`nsToolkitProfileService::Init()` calls](https://searchfox.org/mozilla-central/source/toolkit/profile/nsToolkitProfileService.cpp#792) ↓
* `nsXREDirProvider::GetInstallHash()`:
  * [gets browser executable, then its parent directory](https://searchfox.org/mozilla-central/source/toolkit/xre/nsXREDirProvider.cpp#1195)
  * [calls](https://searchfox.org/mozilla-central/source/toolkit/xre/nsXREDirProvider.cpp#1225) ↓
* [`HashInstallPath` calls](https://searchfox.org/mozilla-central/source/toolkit/xre/nsXREDirProvider.cpp#1173) ↓
* `::GetInstallHash()` [calls `CityHash64()`](https://searchfox.org/mozilla-central/source/toolkit/mozapps/update/common/commonupdatedir.cpp#771)
* Back up the return chain, `nsToolkitProfileService::Init()` [looks up key `Default` in section `Install<hash>`](https://searchfox.org/mozilla-central/source/toolkit/profile/nsToolkitProfileService.cpp#798).

Looks to me like Firefox does indeed re-compute the hash every time it runs, which means there's a good chance it doesn't store it.

**Comment by bovine3dom** [comment]
https://github.com/google/cityhash we could just include it in the native messenger.

If it's fast enough for Firefox it's probably fast enough for us

**Comment by Rummskartoffel** [comment]
Maybe we could move some more native-related code into the native messenger as well, so we don't have to make as many calls to it.

**Comment by bovine3dom** [comment]
The problem is that the more we do that more we have to update the native messenger (not only for the code initially but also to fix bugs).

Updating the native messenger is in theory automatic but it doesn't seem to work in practice. It's much easier to update the extension code.

Keeping responsibility in JS should also help us to keep the native messenger useful to users writing their own scripts. The more we give ourselves special powers the more likely users are going to become second class citizens.

**Comment by Rummskartoffel** [comment]
Good point. I do think we should think of *some* way to limit calls to the messenger, though – for instance, on non-Windows, `:restart` takes between 5 and 9 – if I haven't miscounted –, which I think is too much.

**Comment by bovine3dom** [comment]
We should probably cache `nativegate` https://github.com/tridactyl/tridactyl/blob/cc9db0cd071d5f2b6035bbdef767b3caa6cddbfb/src/lib/native.ts#L229 for a few seconds since that'll be the bulk of the calls, I imagine.

The real solution is to switch to a port based messenger daemon, but that would require a fairly big rewrite of the messenger (again...) and probably a fair bit of extra code in Tridactyl.

**Comment by Rummskartoffel** [comment]
There's a few of those, but I also saw at least two calls to `ff_cmdline` in there. We should be able to cache that indefinitely, since it's not going to change till the browser restarts.

**Comment by bovine3dom** [comment]
I thought we already did :thinking:

**Comment by Rummskartoffel** [comment]
Nope, doesn't look like we do.

**Comment by bovine3dom** [comment]
https://github.com/medikoo/memoizee#user-content-expire-cache-after-given-period-of-time minimal thinking required

**Comment by Rummskartoffel** [comment]
Sounds like a great opportunity to move our existing caching to that and get rid of globals.

On another note, to reproduce Firefox's hashes, looks like we'll need to:
1. get the install directory of the current Firefox
2. convert that path to a wide string if isn't one already
3. cast – not convert – that back to a narrow string
4. pass that into `CityHash64`.

**Comment by Rummskartoffel** [comment]
I've played around with CityHash a bit:
* We won't be able to compile it for Windows unmodified, because it tries to include `<byteswap.h>`, which isn't present in mingw-w64. It might be easier to use Mozilla's version instead, but I haven't tried that out yet.
* I've not yet figured out how to do in Nim the type casting required to emulate what Firefox does.
* I haven't had any luck reproducing the hashes yet. [Here's](https://gist.github.com/Rummskartoffel/34e8651990d74df24ccdca01b6e86242) the code I tried it with for my WSL2 install. The target hash is `4F96D1932A9F858E`, I'm getting `46B8023B0FDD1B34`. If you want to try it yourself, from what I can tell, `path` has to be the parent directory of what Firefox would get from `/proc/self/exe`.

**Comment by bovine3dom** [comment]
> We won't be able to compile it for Windows unmodified, because it tries to include `<byteswap.h>`, which isn't present in mingw-w64. It might be easier to use Mozilla's version instead, but I haven't tried that out yet.

No big deal as we can read the hash from the registry right?

I'm slightly confused about why we need this. I thought in OSX and Linux we could find the profile by checking for a lock file. I thought reading profiles.ini was only for Windows.

**Comment by Rummskartoffel** [comment]
If we don't actually need to read the file outside of Windows, we don't need this either, no. It's been a while since I read the profile detection code, so I just assumed we needed to read it on every OS.

**Comment by bovine3dom** [comment]
That code is a warren.

We read profiles.ini to get all the possible profiles and then use the lock to decide which profile to pick: https://github.com/tridactyl/tridactyl/blob/c37337c7ba4bcb2e0641a31f3c84a303dbce7e08/src/lib/native.ts#L625

I think that should continue to work (indeed we've had no complaints).

**Comment by Rummskartoffel** [comment]
That gives me an idea: what if someone has two different Firefoxes running at the same time? Wouldn't that detect the wrong lock?

**Comment by glacambre** [comment]
>what if someone has two different Firefoxes running at the same time? Wouldn't that detect the wrong lock?

No, because (on linux at least) we check the arguments that were used to start the current Firefox instance. If `-P`/`--profile` is found, then we know what profile is being used. If these arguments aren't found and we find multiple lock files, we know we're using the default profile.

**Comment by Rhys-T** [comment]
`getProfile` seems to be returning the wrong profile on my machine, and I think it might be at least partially because of this. I'm using Firefox 104.0 (normal release) on macOS. I've never run anything older than 84.0.1 on this machine, but I still have both `default` (unused) and `default-release` profiles. (My understanding is that `default` is basically a dummy profile, to keep any old versions that don't know about per-installation profiles from trying to take over `default-release`.)

What seems to happen is:
- `find` doesn't see the `.parentlock` file, because it's starting at `~/Library/Application Support/Firefox/` rather than `…/Profiles/`, and `-maxdepth 2` doesn't go far enough. (`Profiles` is at depth 1, `abcd1234.default-release` is at 2, and `.parentlock` is at 3.)
- That makes Tridactyl fall back to looking for the profile that has `Default=1`, which is the old-style `default` profile that I'm not using. But even if it did find `.parentlock`, it still wouldn't work, because:
- Even though I'm only actually using `default-release`, Firefox seems to be creating `.parentlock` in _both_ profile directories when it runs. Tridactyl wouldn't see _exactly one_ profile in use, so it would still fall back to looking for `Default=1`. (Tested by modifying `hacky_profile_finder` in the debugger just before it's used.)

**Comment by Rhys-T** [comment]
Actually, I was wrong: it is still creating just the `*.default-release/.parentlock` file under normal circumstances. For some reason, opening `about:profiles` causes `*.default/.parentlock` to be created, and then left there indefinitely. The part about the `find` command's `-maxdepth` is still correct, though.

I did find an alternate way of locating the correct profile on *nix systems that might work, however: use `lsof` to find a `.parentlock` file that Firefox actually has open. Basically I'm doing:
```sh
lsof -Fn -p$ppid 2>&1 | grep '^n.*/\.parentlock$' | head -n1
```
where $ppid is Firefox's PID, determined the same way that `ff_cmdline()` does it. That gives me the path to the `.parentlock` that _this instance_ of Firefox has open, prefixed by an `n` character (because of the way that `lsof`'s machine-readable output is structured) and followed by a newline. This only ever returns the actual profile's lock, even when `about:profiles` creates the other one. (It only ever seems to return one result - the `head -n1` is just to make sure. The `2>&1` is to keep any warnings from `lsof` from confusing things.) I can also generate the profile path directly (though still with a newline) using:
```sh
lsof -Fn -p$ppid 2>&1 | sed -En '/^n(.*)\/\.parentlock$/ { s//\1/; p; q; }'
```

I've only tested these on macOS so far, and I don't know whether `lsof` is usually installed on common Linux and *BSD systems or not, but if it is there, it should be more reliable than the current approach. This is also likely to break if the profile path contains newlines or other weird characters, but that's unlikely to happen on a normal system.
